src/controllers/crate_owner_invitation.rs

```diff
@@ -83,6 +55,200 @@ pub fn list(req: &mut dyn RequestExt) -> EndpointResult {
     }))
 }
 
+/// Handles the `GET /api/private/crate-owner-invitations` route.
+pub fn private_list(req: &mut dyn RequestExt) -> EndpointResult {
+    let auth = req.authenticate()?.forbid_api_token_auth()?;
+
+    let filter = if let Some(crate_name) = req.query().get("crate_name") {
+        ListFilter::CrateName(crate_name.clone())
+    } else if let Some(id) = req.query().get("invitee_id").and_then(|i| i.parse().ok()) {
+        ListFilter::InviteeId(id)
+    } else {
+        return Err(bad_request("missing or invalid filter"));
+    };
+
+    let list = prepare_list(req, auth, filter)?;
+    Ok(req.json(&list))
+}
+
+enum ListFilter {
+    CrateName(String),
+    InviteeId(i32),
+}
+
+fn prepare_list(
+    req: &mut dyn RequestExt,
+    auth: AuthenticatedUser,
+    filter: ListFilter,
+) -> AppResult<PrivateListResponse> {
+    let pagination: PaginationOptions = PaginationOptions::builder()
+        .enable_pages(false)
+        .enable_seek(true)
+        .gather(req)?;
+
+    let user = auth.user();
+    let conn = req.db_read_only()?;
+    let config = &req.app().config;
+
+    let mut crate_names = HashMap::new();
+    let mut users = IndexMap::new();
+    users.insert(user.id, user.clone());
+
+    let sql_filter: Box<dyn BoxableExpression<crate_owner_invitations::table, Pg, SqlType = Bool>> =
+        match filter {
+            ListFilter::CrateName(crate_name) => {
+                // Only allow crate owners to query pending invitations for their crate.
+                let krate: Crate = Crate::by_name(&crate_name).first(&*conn)?;
+                let owners = krate.owners(&*conn)?;
+                if user.rights(req.app(), &owners)? != Rights::Full {
+                    return Err(forbidden());
+                }
+
+                // Cache the crate name to avoid querying it from the database again
+                crate_names.insert(krate.id, krate.name.clone());
+
+                Box::new(crate_owner_invitations::crate_id.eq(krate.id))
```

Comment:
is there a reason why we need to box these? wouldn't it be easier to do something like in the `search` controller, where we prepare a generic `query` and then do `query = query.filter(...)` in the condition branches?

Reply:
The search controller [also uses boxes](https://github.com/rust-lang/crates.io/blob/3a01128b314fa67063f22857967ba58ce0b5cf88/src/controllers/krate/search.rs#L58).